.githooks/pre-commit

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env bash
+#
+# Pre-commit hook to check if CHANGELOG.md has been included
+# when the project root or relevant module has been edited.
+#
+# If it has been included, the latest version is compared
+# against the previous version by stripping all leading zeros
+# and comparing them. If it is greater it goes on to check
+# that the latest change's date is the current date (i.e. today).
+# If any of these tests fail the script exits and cancels the
+# commit.
+#
+# To bypass this validation add the `--no-verify` or `-n` flag:
+# 		`git commit --no-verify ...`
+
+# Get names of staged files being committed. If being run as
+# a git hook, take the result of git diff. If not, receive
+# the edited files from the CLI arguments provided.
+if [[ "${#}" -eq 0 ]]
+then
+    STAGED_FILES=($(git diff --cached --name-only)) # Local git hook
+else
+    STAGED_FILES=("${@}") # Github workflow passes arguments
+fi
+
+if [[ "${#STAGED_FILES[@]}" == 0 ]]
+then
+    echo "No files modified"
+    exit 0
+fi
+
+ROOT_EDITED=0
+ROOT_CHANGELOG_INCLUDED=0
+MODULES_EDITED=()
+CHANGELOG_ERRORS_FOUND=0
+IGNORE_DIRS=(".github" ".githooks" "docs" "bin")
+
+# Loop through staged files
+for file in "${STAGED_FILES[@]}"
+do
+    # Check if files are in root directory
+    if [[ "${file}" =~ /+ ]]
+    then
+        root=0
+    else
+        ROOT_EDITED=1
+        root=1
+    fi
+    # Check for CHANGELOG.md in root directory
+    if [[ "${root}" == 1 && "${file}" =~ CHANGELOG.md ]]
+    then
+        ROOT_CHANGELOG_INCLUDED=1
+    fi
+    if [[ "${root}" == 0 ]]
+    then
+        module="$(echo "${file}" | sed 's/\/.*//')"
+        # Only add to modules list if not already present
+        if [[ ! "${MODULES_EDITED[*]}" =~ ${module} ]]
+        then
+            [[ ! "${IGNORE_DIRS[*]}" =~ ${module} ]] && MODULES_EDITED+=("${module}")
+        fi
+    fi
+done
+
+# Check if the root directory has been changed and if the
+# root CHANGELOG.md file was included in the commit
+if [[ "${ROOT_EDITED}" == 1 && "${ROOT_CHANGELOG_INCLUDED}" == 0 ]]
+then
+    printf 'Root changelog not included\n\n' >&2
+    CHANGELOG_ERRORS_FOUND=1
+fi
+
+# Check that for each module edited the CHANGELOG.md file
+# was included in the commit
+MODULES_MISSING_CHANGELOG=()
+CHANGELOG_FILES=() # Store included changelog locations
+if [[ "${#MODULES_EDITED[@]}" != 0 ]]
+then
+    # Loop through edited modules checking for CHANGELOG.md
+    for module in "${MODULES_EDITED[@]}"
+    do
+        MODULE_CHANGELOG_INCLUDED=0
+        for file in "${STAGED_FILES[@]}"
+        do
+            if [[ "${file}" =~ "${module}"/* ]]
```

Comment:
Is this correct?

```suggestion
            # If any files under /root/module/* were edited, we need to make sure that /root/module/CHANGELOG was updated
            if [[ "${file}" =~ "${module}"/* ]]
```

Reply:
Yes this is correct `=~` basically means do a RegEx comparison of the two strings. So this checks only the files for the module currently being checked, and is looking for the `CHANGELOG.md` where ever it may be in the module

Reply:
Appreciate the explanation!